src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -60,85 +47,62 @@ internal bool TryGetBuffer(in SequencePosition start, in SequencePosition end, o
 
                 data = new ReadOnlyMemory<T>(Unsafe.As<T[]>(startObject));
             }
-            else if (type == SequenceType.MemoryManager)
-            {
-                Debug.Assert(startObject is MemoryManager<T>);
-
-                data = (Unsafe.As<MemoryManager<T>>(startObject)).Memory;
-            }
             else if (typeof(T) == typeof(char) && type == SequenceType.String)
             {
                 Debug.Assert(startObject is string);
 
                 data = (ReadOnlyMemory<T>)(object)(Unsafe.As<string>(startObject)).AsMemory();
             }
-            else
+            else // if (type == SequenceType.MemoryManager)
             {
-                data = default;
-                return false;
+                Debug.Assert(startObject is MemoryManager<T>);
+
+                data = (Unsafe.As<MemoryManager<T>>(startObject)).Memory;
             }
 
-            data = data.Slice(startIndex, length);
+            data = data.Slice(startIndex, endIndex - startIndex);
             return true;
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal ReadOnlyMemory<T> GetFirstBuffer(in SequencePosition start, in SequencePosition end)
+        private ReadOnlyMemory<T> GetFirstBuffer(in SequencePosition start, in SequencePosition end)
         {
-            SequenceType type;
-            int startIndex = 0;
-            int endIndex = 0;
+            GetTypeAndIndices(start.GetInteger(), end.GetInteger(), out SequenceType type, out int startIndex, out int endIndex);
             object startObject = start.GetObject();
-            if (startObject != null)
-            {
-                GetTypeAndIndices(start.GetInteger(), end.GetInteger(), out type, out startIndex, out endIndex);
-            }
-            else
-            {
-                type = SequenceType.Empty;
-            }
-
-            int length = endIndex - startIndex;
-            object endObject = end.GetObject();
-
-            if (type != SequenceType.MultiSegment && type != SequenceType.Empty && startObject != endObject)
-                ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
```

Comment:
Now that we removed this check, do we still use this ThrowInvalidOperationException_EndPositionNotReached anywhere? If not, can we remove it from ThrowHelper?

Reply:
`ThrowInvalidOperationException_EndPositionNotReached` is used in `CheckEndReachable` method also. This method is slow and I will remove it later when I will create performance tests for GetPosition and Slice methods and I will suggest more performance code for Seek method.